WIP on the first idea

jviereck opened the pull request.

jviereck committed:
WIP on the first idea

jviereck committed:
Fix react-style link to /simple

SanderSpies commented on 01_basic_idea/README.md:
```diff
@@ -7,3 +7,94 @@ https://github.com/js-next/react-style/tree/f957e09dfb4948/examples/0.13/syntax
+  props.active ? ButtonStyles.activeStyle : null,
+  state.hover ? ButtonStyles.hoverStyle : null,
+  state.focus ? ButtonStyles.focusStyle : null
+].concat(props.styles);
```
The next version of React Style will have pseudo classes and should make this code smaller.

jviereck replied:
Thanks for that hint. Out of interest: How will you support pseudo classes in React Styles - using JavaScript to detect the hover and then update the styles or do you plan to emit a piece of CSS to use the browser `:hover { ... }` selector?

jviereck committed:
Introduce explicit :BASE: rule and introduce StyleSheet.composer function

jviereck committed:
Update the README.md to reflect more the latest changes.

jviereck commented:
Just pushed a few more changes. These changes move the code further away from `react-style`. I tried to come up with a good way to solve the problem of extending existing styles like in

``` js
<Button styles={[ButtonStyles.error, ApplicationStyles.childStyle]}>
```

and ended up introducing what I call "composer" functions, such that the above becomes now:

``` js
<Button styleDef={ButtonStyles.error(ApplicationStyles.childStyle)]}>
```

The points 3) onwards in the [README.md](https://github.com/VirtualCSS/planning/blob/bc325902374f8ab48c31d45868d35d5e99d28ed3/01_basic_idea/README.md) should hopefully give you and idea about the changes and the last point 7) covers the idea behind the composer functions.

The idea about composer functions is highly inspired by the blog post [Mixins Are Dead. Long Live Composition](https://medium.com/@dan_abramov/mixins-are-dead-long-live-higher-order-components-94a0d2f9e750), which talks about the trend in React to move away from mixins and towards composition, and the linearization concept in type systems like Scala. In fact, many of the latest changes came to my mind when thinking about extending CSS classes as a problem in other object oriented programming languages.

As always, if you have any questions or want me to explain details in more depth, please let me know. 

---

Do you think the idea about the "composer functions" is too far fetched or do you think they actually make sense? I personally like the idea very much but I am not sure if my love for type systems makes me biased here.

jviereck commented:
> As always, if you have any questions or want me to explain details in more depth, please let me know. 

Aka, if you want, I am happy to write up more about the composer idea and turn it into a blog post :)

jviereck committed:
Remove way to specify inheritance and rename :BASE: to !BASE

jviereck commented:
Two small changes that I've just pushed:
- renamed `:BASE:` to `!BASE`: This is much easier to read (because there are no two colons at the end of the JSON definition)
- removed the way to specify the inheritance explicit (before done via `:INHERIT-PARENT:`). This is as this feature complicates the design and for now I want to focus on a simple solution that is easier to prototype first.

jviereck committed:
Adding minimal package.json to keep track of dev dependencies

jviereck committed:
Got basic webpack setup working

jviereck committed:
Fixing typos and reshape code to make analysis easier later

jviereck committed:
WIP: got .create implemented but not tested yet

jviereck committed:
WIP: Finished the .compose and .composer function

jviereck committed:
WIP: Got basic CSS mounted and system up and wronging. Something seems wrong with composition though

jviereck committed:
WIP: Generalize deepAssign function

jviereck committed:
WIP: Fix deepAssign

jviereck committed:
Include normalize.css - finished first runtime VirtualCSS implementation

jviereck committed:
Adding missing virtual-css/buildCSSRule.js
